_posts/2024-08-14-running-a-hsm-realm.md

```diff
@@ -0,0 +1,221 @@
+
+## Load Balancers
+
+Clients send their requests to the realm over HTTPS to the load balancer
+instances. The load balancer will authenticate the requests JWT token and then
+route it to the relevant agent. You'll need TLS keys and certificates to start
+the load balancer service. In addition you'll need an ingress service to route
+traffic across the load balancer service instances. If you're using relatively
+short lived TLS keys/certificates such as those generated by [Lets
+Encrypt](https://letsencrypt.org), you can send the HUP signal to the load
+balancer process to reload the key & certificate without impacting inflight
+requests.
+
+The load balancers expects to find a secret in GCP Secret Manager called
+`record-id-randomization` that contains a 32 byte key. This key is used to
+randomize the hash used to generate a merkle tree record id from the tenant and
+user information. If you have the Google `gcloud` tools installed then this
+snippet can be used to create the key.
+
+```sh
+python3 -c \
+'import secrets, json; print(json.dumps({"data": secrets.token_hex(32), "encoding": "Hex", "algorithm": "Blake2sMac256"}), end="")' | \
+    gcloud \
+        --project <gcp-project> \
+        secrets create \
+        record-id-randomization \
+        --data-file - \
+        --labels kind=record_id_randomization_key
+```
+
+## Realm Initialization
+
+Once the agent is successfully running on each HSM server the realm
+initialization can be performed. The
+[`cluster`](https://github.com/juicebox-systems/juicebox-hsm-realm/tree/main/cluster_cli)
+tool is used to perform many administrative tasks. The `cluster agents` command
+can be used to sanity check that the agents are running and accessible.
+
+First use `cluster new-realm <some-agent>` to crete a new realm. This will
+assign the realm its IDs and create and persist the initial merkle tree.
+
+Then use `cluster join-realm --realm <realmID> <agent1> <agent2> ..` to have the
+other HSMs join the same realm.
+
+The initial realm creation created a single replication group with a single
+member. At this point new replication groups can be created (via `cluster
+new-group`) and the groups given ownership of part of the key space (via
+`cluster transfer`). The `cluster groups` command can give a detailed report of
+each replication groups status.
+
+You'll need to decide the exact number of replication groups, and the
+distribution of their members. Differing sizes of clusters may use different
+approaches. For a 5 HSM cluster we recommend 3 replication groups each
+containing all 5 HSMs as members and each replication group owning a third of
+the key space. In larger clusters its not required that all HSMs are members of
+all replication groups.
+
+## Services
+
+The agent, load balancer and cluster manager services will all panic in
+unexpected and unrecoverable error conditions. A process supervisor such as
+Systemd should be used to manage these services and restart the process as
```

Comment:
Can we include example service files (since we have them in the infra repo?) or is that too much you think?

Reply:
The infra repo ones are templated based on some of our weird mixed environment requirements, so aren't directly applicable. And we've already blown alway all the expanded versions. I'm okay with them having to put their own together. If that's a blocker for them they've got much larger problems.